accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,341 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+  - If the variable is declared `final` and not `late` it is an error, and there
+    is no **initialization inferred type** after the join.
+  - If the variable is declared `late` and the **initialization inferred type**
+    on each path which has one is the same type `T` (using syntactic equality),
+    then the **initialization inferred type** after the join is `T`.
+  - If the variable is declared `late` and/or `var` and the **initialization
+    inferred type** on one or more paths is the same type `T` (using syntactic
+    equality), and the **initialization inferred type** on one or more paths is
+    `Null`, then the **initialization inferred type** after the join is `T?`.
+  - If the variable is declared `var` and not `late` and the **initialization
+    inferred type** on one or more paths is the same type `T` (using syntactic
+    equality), and the **initialization inferred type** on all other paths (if
+    any) is `Null`, then the **initialization inferred type** after the join is
+    `T?`.
+  - Otherwise it is an error.
+
+In other words, a non-`late`, `final` **untyped** variable must be assigned a
+value on every path, but a variable declared `late` and/or `var` may be assigned
+a value on only some of the paths into a join.  In the latter case, the type is
+made nullable if the variable is declared as non-`late`, and the variable is
+treated as implicitly null initialized.  Variables declared `late` are only made
+nullable if `Null` is explicitly assigned on one more paths into the join, and
+there is no implicit null initialization.  In all cases, the values assigned
+must all have the same type, or type `Null`.
+
+
+```dart
+void test(bool b) {
+  { // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    } else {
+      x = 3.0;
+    } // Error: x has two different types on different paths
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    } else {
+      x = null;
+    } // x has type int?
+  }
+
+  {
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // x has type int?
+  }
+
+  {
+    final x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // Error: not assigned on all paths
+  }
+
+  { // late final the same
+    late x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // x has type int
+  }
+
+  var z; // z is an untyped variable
+  if (b) {
+    z = 3; // z is promoted to int
+   } else {
+    z = 3.0; // z is promoted to double
+    return;
+  } // No error.  z is inferred as int on the only reachable path
+  z.isEven; // No error, only one reachable path
+
+  var u; // u is an untyped variable
+  if (b) {
+    u = 0 as num; // u is promoted to num
+    if (u is! int) u = 1;
+    // u has inferred type num, but is promoted to int
+  } else {
+    u = 3;
+  } // Error: u is inferred to different types on different paths
```

Comment:
Subtle! This will trip people up when they run into it. Is it a crazy idea to use the promoted type at a join point instead of just the inferred type?

Reply:
Alternatively, we can just make it have *no* inferred initialization type, and make it an error to read or write a variable with no type. The variable will still be potentially assigned so no further initialization can happen.
Then
```dart
foo() {
  var x;
  if (test) {
    x = 4;
    something(x);
  } else if (test2) {
    x = 4.0;
    somethingElse(x);
  } else {
    throw "whatever";
  }
  return;
}
```
would be valid because even though there is no type for `x` after the if-chain, there is also no use of it.

it's different because we have to let the program survive in a state where a variable has no type, rather than make it an error at the join point, which is why we have to explicitly call out that the variable is unusable afterwards.

Reply:
> This will trip people up when they run into it. Is it a crazy idea to use the promoted type at a join point instead of just the inferred type?

Not sure, I can think about it.

Reply:
> Alternatively, we can just make it have _no_ inferred initialization type, and make it an error to read or write a variable with no type.

It is already an error to read a variable with no IIT.  We can make it an error to write, or not.

The general idea of making the IIT non-monotonic feels icky, but maybe would work out fine.  You'd have a variable which was in a usable state for some period of time, and then becomes unusable.  Feels weird, but probably doable.

Reply:
> Subtle! This will trip people up when they run into it. Is it a crazy idea to use the promoted type at a join point instead of just the inferred type?

Where this starts getting complicated is deciding what combinations you allow.  If I infer a variable as `Object` on one path, then promoted it from there to `num`, and then to `int`, what joins do I allow?  Can I join it with another path that inferred `Object` (demoting to `Object`)?  Can I join it with another path that inferred `num` (demoting to `num`)?  What about another path that inferred `num?`?  Do I have to look at every pair of types in the promotion chain, with their nullable + non-nullable versions?  And what about when there are multiple paths into the join?  I'll think about whether there's a clean story here, but it feels really icky on first glance.